File: packages/gbnf-experiment/data/20260909T013603Z_054a84ca/ported_implementation/gbnf/rules_builder/symbol_ids.py
from __future__ import annotations

from typing import Dict, Iterator, KeysView, Tuple


class SymbolIds:
    """Bidirectional name <-> id map, also recording where a name was defined."""

    # we don't need to delete, just preserve relationships
    def __init__(self) -> None:
        self._map: Dict[str, int] = {}
        self._pos: Dict[str, int] = {}
        self._reverse_map: Dict[int, str] = {}

    @property
    def size(self) -> int:
        return len(self._map)

    def keys(self) -> KeysView[str]:
        return self._map.keys()

    def has(self, key: str) -> bool:
        return key in self._map

    def get(self, key: str) -> int:
        if key not in self._map:
            raise KeyError(f'SymbolIds does not contain key: {key}')
        return self._map[key]

    def reverse_get(self, key: int) -> str:
        if key not in self._reverse_map:
            raise KeyError(f'SymbolIds does not contain value: {key}')
        return self._reverse_map[key]

    def get_pos(self, key: str) -> int:
        if key not in self._pos:
            raise KeyError(f'SymbolIds does not contain key: {key}')
        return self._pos[key]

    def set(self, key: str, value: int, pos: int) -> None:
        self._map[key] = value
        self._pos[key] = pos
        self._reverse_map[value] = key

    def __len__(self) -> int:
        return len(self._map)

    def __iter__(self) -> Iterator[Tuple[str, int]]:
        return iter(self._map.items())

File: proposed-projects/round-trip-experiment/reverse/20260909T140517Z_670261ce/20260910T081441Z_4899d5cf/ported_implementation/src/rules-builder/symbol-ids.ts
// we don't need to delete, just preserve relationships
export class SymbolIds {
  private readonly map = new Map<string, number>();
  private readonly pos = new Map<string, number>();
  private readonly reverseMap = new Map<number, string>();

  get size(): number {
    return this.map.size;
  }

  keys(): IterableIterator<string> {
    return this.map.keys();
  }

  has(key: string): boolean {
    return this.map.has(key);
  }

  get(key: string): number {
    const val = this.map.get(key);
    if (val === undefined) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return val;
  }

  reverseGet(key: number): string {
    const val = this.reverseMap.get(key);
    if (val === undefined) {
      throw new Error(`SymbolIds does not contain value: ${key}`);
    }
    return val;
  }

  getPos(key: string): number {
    const pos = this.pos.get(key);
    if (pos === undefined) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return pos;
  }

  set(key: string, value: number, pos: number): void {
    this.map.set(key, value);
    this.pos.set(key, pos);
    this.reverseMap.set(value, key);
  }

  [Symbol.iterator](): IterableIterator<[string, number]> {
    return this.map.entries();
  }
}
